### PR TITLE
feat: update election post to use flexible Tree structure

### PR DESCRIPTION
**Pull request by cryptonemo**
Side note: was planning to watch the CI process on this PR and may put add related updates here

### CODE REVIEW

File: storage-proofs/src/post/election/circuit.rs [hunk 1]
```diff
@@ -137,7 +135,7 @@ impl<'a, H: 'static + Hasher> Circuit<Bls12> for ElectionPoStCircuit<H> {
         }
 
         // pad to a multiple of md arity
-        let arity = PoseidonMDArity::to_usize();
+        let arity = Tree::Arity::to_usize();
```
Comment (dignifiedquire) [inline storage-proofs/src/post/election/circuit.rs:138]: this should not be changed, this is a fixed arity that depends on the number of challenges, not the tree

File: storage-proofs/src/post/election/circuit.rs [hunk 2]
```diff
@@ -146,7 +144,7 @@ impl<'a, H: 'static + Hasher> Circuit<Bls12> for ElectionPoStCircuit<H> {
         }
 
         // hash it
-        let partial_ticket_num = PoseidonFunction::hash_md_circuit::<_>(
+        let partial_ticket_num = <Tree::Hasher as Hasher>::Function::hash_md_circuit::<_>(
```
Comment (dignifiedquire) [inline storage-proofs/src/post/election/circuit.rs:147]: This should stay hardcoded to Poseidon

File: storage-proofs/src/post/election/vanilla.rs [hunk 3]
```diff
@@ -198,13 +218,13 @@ fn generate_candidate<H: Hasher>(
     }
 
     // pad for md
-    let arity = PoseidonMDArity::to_usize();
+    let arity = Tree::Arity::to_usize();
```
Comment (dignifiedquire) [inline storage-proofs/src/post/election/vanilla.rs:221]: should stay at poseidonmd

File: storage-proofs/src/post/election/vanilla.rs [hunk 4]
```diff
@@ -198,13 +218,13 @@ fn generate_candidate<H: Hasher>(
...
     while data.len() % arity != 0 {
-        data.push(PoseidonDomain::default());
+        data.push(<Tree::Hasher as Hasher>::Domain::default());
```
Comment (dignifiedquire) [inline storage-proofs/src/post/election/vanilla.rs:223]: keep as poseidon

File: storage-proofs/src/post/election/vanilla.rs [hunk 5]
```diff
@@ -198,13 +218,13 @@ fn generate_candidate<H: Hasher>(
...
     }
 
     let partial_ticket: Fr = measure_op(Operation::PostPartialTicketHash, || {
-        PoseidonFunction::hash_md(&data)
+        <Tree::Hasher as Hasher>::Function::hash_md(&data)
```
Comment (dignifiedquire) [inline storage-proofs/src/post/election/vanilla.rs:227]: keep as poseidon

**Comment by cryptonemo** [comment]
> please add at least on test for sub and one for top trees, so we know those work as well

Updated

**Pull request merged**